text/dependencies/rfcs/0000-dependency-management-phase-one.md

```diff
@@ -0,0 +1,416 @@
+| tini              | https://github.com/krallin/tini/tarball                   | yes, buildpacks-ci                       |
+| yarn              | https://github.com/yarnpkg/yarn/releases/download/v1.15.2 | processed, buildpacks-ci                 |
+</details>
+
+### Transfer Compilation Code
+
+For any dependencies that must still be compiled or processed in some way, the
+code from the Cloud Foundry repositories should be rewritten, and moved over to
+the buildpack under a directory called `dependency/compilation`. The code
+should be easily runnable from a local machine to compile the code, and should
+be documented with a README about how to use it. As outlined in the
+[Overarching
+Plan](https://github.com/paketo-buildpacks/rfcs/blob/dependency-management-step-one/text/dependencies/rfcs/0000-dependency-management-phase-one.md#overarching-plan)
+section, it will take in the `version`, the `outputDir` to put the dependency,
+and the `target` image to compile against.
+
+The code will eventually be used in a Github Actions workflow, so the location,
+inputs, and outputs needs to be standardized across buildpacks. The workflows
+and Github Actions to use this code will be enumerated in a separate RFC.
+
+The dependency will likely need to be compiled against different
+stacks/architectures, so this should also be enumerated in the RFC and taken
+account of in the new compilation code. This should include information about
+the different images we will need to compile against, as well as the stacks
+that will be compatible with that dependency. Maintainers should also consider
+if the dependency of interest is OS distribution-agnostic, or if it will need
+to be compiled separately depending on the distribution or platform it's used
+on.
+
+##### Specifiy variants with the `targets.json` file
+A file under the path `<buildpack>/dependency/targets.json` will be an
+enumeration of the different variants of a dependency, based on the different
+stacks the buildpack supports. This will be parsed and used by workflows
+described in Phase 2, but also serves as a manifest of the different version
+variants of each dependency.
+```
+[
+  {
+    "target": <generic name of the group, to be used in tarball naming if compiled>,
+    "image": <image to compile against>,
+    "stacks": [ <list of compatible stacks>]
+  },
+  ...
+]
+```
+
+The images listed in the `target.json` file will directly be used as the image
+used during the compilation step during Github Action runs. When compiling a
+dependency locally, users should be able to pull the image themselves and run
+the compilation code on the container.
+
+### Image Preparation
+In the event that a dependency is compiled as outlined in the section above,
+the compilation will take place on a Docker image of the buildpack maintainers'
+choosing. Hopefully, the image that will be used will contain everything needed
```

Comment:
Is there a way for maintainers to define a Dockerfile for the desired compilation environment – without needing to store that image in a registry somewhere?

Reply:
We looked into this by way of having compilation be written as an action, with different Dockerfiles for different compilation environment possibilities. GHA environments don't allow for dynamic actions like that so we couldn't do it like that really. This is where the preparation script came from, it'll essentially take an existent image and stuff on top of it if needed.

We figured in most cases, the environment can be an existent stack image, in which case adding things won't be necessary anyway.

Reply:
Maintainers could also maintain a docker image completely outside the scope of dependencies and then reference that docker image for the dependencies workflow. I know it's a few extra steps but we couldn't find a way for the workflow to allow for buildpack authors to choose either an image or a Dockerfile, so we chose to use images.

Reply:
The workflow will allow buildpack authors to specify some additional "preparation steps" in case an off-the-shelf image doesn't quite have everything. Those steps will be in the form of a shell script that will only run for the target. Something like so:

```shell
If target=="macos";
  brew update && brew install coreutils
else if target=="ubuntu";
  apt-get update && apt-get install -y wget
else if target=="alpine";
  apk update && apk add wget
fi
```

Reply:
And.... just realized @sophiewigmore 's already described this over under '# Image Preparation'. :)

Reply:
Update from team share out - I am going to look into an alternative to this proposal involving Dockerfiles. Please hold!

Reply:
Alright, per @fg-j  recommendation I tried out an approach of rolling compilation into an action so that the image compiled on is configurable via Dockerfile, so there can be different Dockerfiles for each target. This works **beautifully** locally, allowing users to run compilation locally with a few Docker commands.

Unfortunately, any way you slice it, there's absolutely no way I have discovered to dynamically swap out the Dockerfile used in an action, so combining this strategy with a matrix (i.e. we run compilation for each target) isn't feasible. The only way around it would be if the workflow was hardcoded to run compilation for each target.

Reply:
Check out https://github.com/joshuatcasey/bundler/commit/5009e1635ce5bb27cd77ed64c35ed5375fc28c33 to see what this would look like in practice

Reply:
Alright, the RFC has been updated to reflect this idea as if https://github.com/paketo-buildpacks/rfcs/pull/205/commits/ca21bfc023a52cd51f6e768e65605a64e3d62dd1